### PR TITLE
[SPARK-28890][SQL] Upgrade Hive Metastore Client to the 3.1.2 for Hive 3.1

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Hive 3.1.2 has been released. This PR upgrades the Hive Metastore Client to 3.1.2 for Hive 3.1.

Hive 3.1.2 release notes:
https://issues.apache.org/jira/secure/ReleaseNote.jspa?version=12344397&styleName=Html&projectId=12310843

### Why are the changes needed?

This is an improvement to support a newly release 3.1.2. Otherwise, it will throws `UnsupportedOperationException` if user `set spark.sql.hive.metastore.version=3.1.2`:
```scala
Exception in thread "main" java.lang.UnsupportedOperationException: Unsupported Hive Metastore version (3.1.2). Please set spark.sql.hive.metastore.version with a valid version.
	at org.apache.spark.sql.hive.client.IsolatedClientLoader$.hiveVersion(IsolatedClientLoader.scala:109)
```

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Existing UT

### CODE REVIEW

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala [hunk 1]
```diff
@@ -103,7 +103,7 @@ private[hive] object IsolatedClientLoader extends Logging {
     case "2.2" | "2.2.0" => hive.v2_2
     case "2.3" | "2.3.0" | "2.3.1" | "2.3.2" | "2.3.3" | "2.3.4" | "2.3.5" | "2.3.6" => hive.v2_3
     case "3.0" | "3.0.0" => hive.v3_0
-    case "3.1" | "3.1.0" | "3.1.1" => hive.v3_1
+    case "3.1" | "3.1.0" | "3.1.1" | "3.1.2" => hive.v3_1
```
Comment (HyukjinKwon) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: Shall we update:

https://github.com/apache/spark/blob/02a0cdea13a5eebd27649a60d981de35156ba52c/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala#L64-L67

and https://github.com/apache/spark/blob/02a0cdea13a5eebd27649a60d981de35156ba52c/docs/sql-data-sources-hive-tables.md#interacting-with-different-versions-of-hive-metastore 
?

Comment (wangyum) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala]: Oh, sorry. We also need to update https://github.com/apache/spark/blob/master/docs/sql-migration-guide-hive-compatibility.md

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
cc @dongjoon-hyun

**Review by srowen — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @wangyum , @HyukjinKwon , @srowen !

**Pull request closed without merging**